File: Leetcode/basic-calculator-ii.cpp
// problem link
// https://leetcode.com/problems/basic-calculator-ii/description/

#include<bits/stdc++.h>
using namespace std;

#define ll long long
class Solution {
public:
    bool is_num(char ch) {
        return ch >= '0' && ch <= '9';
    }

    int calculate(string s) {
        stack<ll> st;
        ll num = 0;
        char prev_op = '+';
        for(int i=0; i<s.size(); i++) {
            if(s[i] == ' ') continue;

            if(is_num(s[i])) {
                num = num * 10 + s[i] - '0';
            } else {
                if(!st.empty()) {
                    ll prev = st.top();
                    if(prev_op == '/') {
                        st.pop();
                        num = prev / num;
                    }
                    else if(prev_op == '*') {
                        st.pop();
                        num = prev * num;
                    }
                    else if(prev_op == '-') {
                        num = -num;
                    }
                }
                prev_op = s[i];
                st.push(num);
                num = 0;
            }
        }
        if(!st.empty()) {
            ll prev = st.top();
            if(prev_op == '/') {
                st.pop();
                num = prev / num;
            }
            else if(prev_op == '*') {
                st.pop();
                num = prev * num;
            }
            else if(prev_op == '-') {
                num = -num;
            }
        }
        st.push(num);

        int ans = 0;
        while(!st.empty()) {
            ans += st.top();
            st.pop();
        }

        return ans;
    }
};

File: Leetcode/basic-calculator-ii.js
// problem link
// https://leetcode.com/problems/basic-calculator-ii/description/

const OPERATOR_MAP = new Map([
    ["+", 1],
    ["-", 1],
    ["*", 1],
    ["/", 1],
])

const handle = (a,b, op, st) => {
    switch(op) {
        case "+":
            st.push(a, b);
            break;
        case "-":
            st.push(a, -b);
            break;
        case "*":
            st.push(a*b)
            break;
        case "/":
            st.push(Math.trunc(a/b));
            break;
    }
}

var calculate = function(s) {
    let right = "";
    let op = "*";
    const st = [1]
    const n = s.length;

    for(let i=0; i<n; i++) {
        if(s[i] === " ") continue;

        // is number
        if(OPERATOR_MAP.has(s[i]) === false) {
            right += s[i];
            continue;
        }
        
        // is operator
        handle(st.pop(), Number(right), op, st)
        op = s[i];
        right = "";
    }
    handle(st.pop(), Number(right), op, st)

    let sum = st.reduce((acc, cur) => {
        return acc + cur;
    }, 0);

    return sum;
};